src/mono/mono/metadata/marshal.c

```diff
@@ -5226,39 +5226,129 @@ mono_marshal_get_unsafe_accessor_wrapper (MonoMethod *accessor_method, MonoUnsaf
+		// TODO: this always tries to compile a generic version of the accessor method and
+		// then inflate it.  But maybe we dont' want to do that (particularly for field
+		// accessors).  In particular if there is no method_inst, we're looking at an
+		// accessor method inside a generic class (alwayst a ginst? or sometimes a gtd?)
+		// In that case we might just want to compile the instance.
+
+		orig_method = accessor_method;
+		ctx = &((MonoMethodInflated*)accessor_method)->context;
+		accessor_method = ((MonoMethodInflated*)accessor_method)->declaring;
+		container = mono_method_get_generic_container (accessor_method);
+		if (!container)
+		    container = mono_class_try_get_generic_container (accessor_method->klass);
+		g_assert (container);
+		// TODO:
+		// in the example below, do we need to mess with the context and container?
+		//
+		// class C<T> {
+		//    public static extern void AccessorMethod<U>(List<T> t, List<U> u);
+		// }
+		//
+		// when we make a wrapper
+		//
+		//    public static extern void wrapper_AccessorMethod<U2>(List<T>, List<U2> u);
+		//
+		// do we need to substitute the new gparams of the wrapper, but leave the
+		// gparams of C<T> unchanged?
+		//
+	}
+
+	// printf("work on: %s (generic = %d, inflated = %d)\n", mono_method_full_name(accessor_method, TRUE), accessor_method->is_generic?1:0, accessor_method->is_inflated?1:0);
+
 	/*
 	 * Check cache
 	 */
-	if (ctx) {
-		cache = NULL;
-		g_assert_not_reached ();
+	if (is_inflated) {
+		cache = get_cache (&((MonoMethodInflated*)orig_method)->owner->wrapper_caches.unsafe_accessor_cache , mono_aligned_addr_hash, NULL);
+		res = check_generic_wrapper_cache (cache, orig_method, orig_method, accessor_method);
+		if (res)
+			return res;
 	} else {
 		cache = get_cache (&mono_method_get_wrapper_cache (accessor_method)->unsafe_accessor_cache, mono_aligned_addr_hash, NULL);
 		if ((res = mono_marshal_find_in_cache (cache, accessor_method)))
 			return res;
 	}
+	// printf ("Cache miss\n");
+	
+	mb = mono_mb_new (accessor_method->klass, accessor_method->name, MONO_WRAPPER_OTHER);
+	if (generic_wrapper) {
+		// If the accessor method was generic, make the wrapper generic, too.
 
-	sig = mono_metadata_signature_dup_full (get_method_image (accessor_method), mono_method_signature_internal (accessor_method));
-	sig->pinvoke = 0;
+		// Load a copy of the generic params of the accessor method
+		mb->method->is_generic = generic_wrapper;
+		container = mono_class_try_get_generic_container (accessor_method->klass);
+		container = mono_metadata_load_generic_params (m_class_get_image (accessor_method->klass), accessor_method->token, container, /*owner:*/mb->method);
+		mono_method_set_generic_container (mb->method, container);
 
-	mb = mono_mb_new (accessor_method->klass, accessor_method->name, MONO_WRAPPER_OTHER);
+		MonoGenericContext inst_ctx = {0,};
+		// FIXME: if is_inflated, do we need to mess with ctx?
+		inst_ctx.method_inst = container->context.method_inst;
```

Comment:
So conceptually if both the class is generic and the `static extern` accessor method is generic, when we make a wrapper we want its signature to use the same exact class gparams as the accessor method (conceptually both the accessor method and the wrapper are members of the class).  But we want it to use its own brand new method gparams.  So we take the original signature of hte access method and replace the method gparams:

```csharp

class G<T> {
   public static extern Tuple<T,U> Accessor<U>(T t, U u);
   // wrapper:
   // public static extern Tuple<T, U2> Accessor<U2>(T t, U2 u);
}
```

so it's as if the wrapper instead of having a gparam named `U` has a gparam named `U2`. And so if we just take signature of the accessor, we have to replace all `U`s by `U2`.  But the `T` is the same